### PR TITLE
[experiment] Use Sets instead of lists in more places

### PR DESCRIPTION
**Pull request by jakebailey**
An old experiment inspired by #55224

### CODE REVIEW

**Comment by jakebailey** [comment]
@typescript-bot perf test this faster

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -25983,20 +25987,20 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
...
+                        (matchedSources ??= new Set()).add(s);
+                        (matchedTargets ??= new Set()).add(t);
                     }
                 }
             }
             return [
-                matchedSources ? filter(sources, t => !contains(matchedSources, t)) : sources,
-                matchedTargets ? filter(targets, t => !contains(matchedTargets, t)) : targets,
+                matchedSources ? filter(sources, t => !matchedSources!.has(t)) : sources,
+                matchedTargets ? filter(targets, t => !matchedTargets!.has(t)) : targets,
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts:26003]: This is so weird - why didn't these always just return `matchedSources ?? sources` and `matchedTargets ?? targets` before?

Comment (DanielRosenwasser) [reply src/compiler/checker.ts:26003]: Oh, these are the ones that *don't match*.

**Comment by jakebailey** [comment]
Dissapointing

**Pull request closed without merging**